### PR TITLE
Implement record based Crucible reference counting

### PR DESCRIPTION
**Pull request by jmpesp**
Crucible volumes are created by layering read-write regions over a hierarchy of read-only resources. Originally only a region snapshot could be used as a read-only resource for a volume. With the introduction of read-only regions (created during the region snapshot replacement process) this is no longer true!

Read-only resources can be used by many volumes, and because of this they need to have a reference count so they can be deleted when they're not referenced anymore. The region_snapshot table uses a `volume_references` column, which counts how many uses there are. The region table does not have this column, and more over a simple integer works for reference counting but does not tell you _what_ volume that use is from. This can be determined (see omdb's validate volume references command) but it's information that is tossed out, as Nexus knows what volumes use what resources! Instead, record what read-only resources a volume uses in a new table.

As part of the schema change to add the new `volume_resource_usage` table, a migration is included that will create the appropriate records for all region snapshots.

In testing, a few bugs were found: the worst being that read-only regions did not have their read_only column set to true. This would be a problem if read-only regions are created, but they're currently only created during region snapshot replacement. To detect if any of these regions were created, find all regions that were allocated for a snapshot volume:

    SELECT id FROM region
    WHERE volume_id IN (SELECT volume_id FROM snapshot);

A similar bug was found in the simulated Crucible agent.

This commit also reverts #6728, enabling region snapshot replacement again - it was disabled due to a lack of read-only region reference counting, so it can be enabled once again.

### CODE REVIEW

**Review by leftwo — COMMENTED** [review]
I may have some more questions after I finish going through `n[exus/tests/integration_tests/volume_management.rs](https://github.com/oxidecomputer/omicron/pull/6805/files#diff-2ef483e9a59180948d56fb5b7dc4f56658dc7aa956ce99498ff6d7804d060b32)` but I thought I should give you what I have so far.

File: schema/crdb/crucible-ref-count-records/up08.sql [hunk 1]
```diff
@@ -0,0 +1,29 @@
+INSERT INTO volume_resource_usage (
+  SELECT
+    (
+      OVERLAY(
+       OVERLAY(
+        MD5(volume.id::TEXT || dataset_id::TEXT || region_id::TEXT || snapshot_id::TEXT || snapshot_addr || volume_references::TEXT)
+        PLACING '4' from 13
```
Comment (smklein) [inline schema/crdb/crucible-ref-count-records/up08.sql:7]: What is this doing? Why '4'? Why `from 13`?

File: schema/crdb/crucible-ref-count-records/up08.sql [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+       )
+       PLACING TO_HEX(volume_references) from 17
```
Comment (smklein) [inline schema/crdb/crucible-ref-count-records/up08.sql:9]: Why `from 17`?

File: nexus/db-queries/src/db/datastore/region.rs [hunk 3]
```diff
@@ -264,7 +264,7 @@ impl DataStore {
                 block_size,
                 blocks_per_extent,
                 extent_count,
-                read_only: false,
+                read_only: maybe_snapshot_id.is_some(),
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/region.rs:267]: Unfortunately yes

File: nexus/db-model/src/volume_resource_usage.rs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+/// true.
+///
+/// Read-only resources can be used by many volumes, and because of this they
+/// need to have a reference count so they can be deleted when they're not
+/// referenced anymore. The region_snapshot table uses a `volume_references`
+/// column, which counts how many uses there are. The region table does not have
+/// this column, and more over a simple integer works for reference counting but
+/// does not tell you _what_ volume that use is from. This can be determined
+/// (see omdb's validate volume references command) but it's information that is
+/// tossed out, as Nexus knows what volumes use what resources! Instead, record
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-model/src/volume_resource_usage.rs:37]: In this context I mean "instead of tossing out information, record it in the usage table", where "tossing out information" means to just increment and decrement the `volume_references` integer in the region snapshot table

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 5]
```diff
@@ -91,19 +95,187 @@ enum VolumeGetError {
...
+                    .eq(Some::<db::model::SqlU16>(address.port().into())),
+            )
+            .filter(region_dsl::read_only.eq(read_only))
+            .select(Region::as_select())
+            .get_result_async::<Region>(conn)
+            .await
+            .optional()
+    }
+
+    async fn read_only_target_to_volume_resource_usage(
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs:229]: Correct - only one non-deleted object per target should exist. The lifetime of these objects should be equal to or longer than the actual Crucible resource.

File: nexus/db-model/src/volume_resource_usage.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (leftwo) [reply nexus/db-model/src/volume_resource_usage.rs:37]: Cool, could we update the comment with that?

[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-model/src/volume_resource_usage.rs]: done in 221751a4d

File: nexus/db-model/src/volume_resource_usage.rs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+impl From<VolumeResourceUsageRecord> for VolumeResourceUsage {
+    fn from(record: VolumeResourceUsageRecord) -> VolumeResourceUsage {
+        match record.usage_type {
+            VolumeResourceUsageType::ReadOnlyRegion => {
+                VolumeResourceUsage::ReadOnlyRegion {
+                    region_id: record
+                        .region_id
+                        .expect("valid read-only region usage record"),
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-model/src/volume_resource_usage.rs]: agreed, that's in 0a45763a2

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 7]
```diff
@@ -91,19 +95,187 @@ enum VolumeGetError {
...
+        )
+        .await?;
+
+        if let Some(region) = maybe_region {
+            return Ok(Some(VolumeResourceUsage::ReadOnlyRegion {
+                region_id: region.id(),
+            }));
+        }
+
+        // If the resource was hard-deleted, or in the off chance that the
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs:277]: It's not expected often, no - only during some race conditions where a resource was deleted before a volume that uses it could be created. In those cases we'll see `CouldNotFindResource` errors, so I'm inclined not to create a separate log message.

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 8]
```diff
@@ -934,52 +1215,237 @@ impl DataStore {
...
-                        .or(region_snapshot_dsl::volume_references.is_null()),
+        let mut regions: Vec<Uuid> =
+            Vec::with_capacity(3 * num_read_write_subvolumes);
+
+        let mut region_snapshots: Vec<RegionSnapshotV3> =
+            Vec::with_capacity(crucible_targets.read_only_targets.len());
+
+        // First, grab read-write regions - they're not shared, but they are
+        // not candidates for deletion if there are region snapshots
+        {
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs]: I think that was a hold-over from a previous version of the code that included some `use ...schema...` entries, and this was meant to scope those down. 00d9f3879 fixes this.

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 9]
```diff
@@ -2409,24 +2956,188 @@ impl DataStore {
...
+                            err.bail_retryable_or_else(e, |e| {
+                                VolumeReplaceSnapshotError::Public(
+                                    public_error_from_diesel(
+                                        e,
+                                        ErrorHandler::Server,
+                                    )
+                                )
+                            })
+                        })?
+                        // XXX be smart enough to .filter the above query
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs]: 412d66564

File: schema/crdb/dbinit.sql [hunk 10]
```diff
@@ -4475,6 +4481,43 @@ CREATE INDEX IF NOT EXISTS lookup_bgp_config_by_bgp_announce_set_id ON omicron.p
 ) WHERE
     time_deleted IS NULL;
 
+CREATE TYPE IF NOT EXISTS omicron.public.volume_resource_usage_type AS ENUM (
+  'read_only_region',
+  'region_snapshot'
+);
+
+CREATE TABLE IF NOT EXISTS omicron.public.volume_resource_usage (
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply schema/crdb/dbinit.sql:4497]: you're right, done in bd2ef84d4

File: nexus/db-model/src/volume_resource_usage.rs [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+    pub volume_id: Uuid,
+
+    pub usage_type: VolumeResourceUsageType,
+
+    pub region_id: Option<Uuid>,
+
+    pub region_snapshot_dataset_id: Option<Uuid>,
+    pub region_snapshot_region_id: Option<Uuid>,
+    pub region_snapshot_snapshot_id: Option<Uuid>,
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-model/src/volume_resource_usage.rs:59]: nice, done in bd2ef84d4

File: nexus/db-queries/src/db/datastore/region.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/region.rs:267]: Note though that currently the only thing in Nexus that creates read-only regions is region snapshot replacement, so this wasn't a bug that was hit anywhere.

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 12]
```diff
@@ -781,17 +1019,33 @@ impl DataStore {
...
+    > {
         use db::schema::dataset::dsl as dataset_dsl;
         use db::schema::region::dsl as region_dsl;
         use db::schema::region_snapshot::dsl;
         use db::schema::volume::dsl as volume_dsl;
+        use db::schema::volume_resource_usage::dsl as ru_dsl;
 
         // Find all regions and datasets
-        region_dsl::region
+        let tuples_superset = region_dsl::region
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs]: changed in 2ff34d5e4

**Comment by jmpesp** [comment]
@smklein @leftwo have a look at 2ff34d5e4, which contains a fix for the region deletion bug I've been talking about.

File: schema/crdb/crucible-ref-count-records/up08.sql (hunk 1 above)
Comment (jmpesp) [reply schema/crdb/crucible-ref-count-records/up08.sql:7]: This code creates a deterministic V4 UUID from the other columns, which [has the shape of](https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random))

```
xxxxxxxx-xxxx-Mxxx-Nxxx-xxxxxxxxxxxx
12345678 9111 1111 1112 222222222333
          012 3456 7890 123456789012
```

where `the first hexadecimal digit in the third group always starts with a 4` means that M = 4, and I put a "random" value (read: volume references) in for N (the variant field). 

But from https://datatracker.ietf.org/doc/html/rfc9562#name-uuid-version-4:

> Alternatively, an implementation MAY choose to randomly generate the exact required number of bits for random_a, random_b, and random_c (122 bits total) and then concatenate the version and variant in the required position.

I don't think I'm doing this right - position 17 shouldn't be "random":

> var:
>     The 2-bit variant field as defined by [Section 4.1](https://datatracker.ietf.org/doc/html/rfc9562#variant_field), set to 0b10.

I can fix this tomorrow

File: schema/crdb/crucible-ref-count-records/up08.sql (hunk 2 above)
Comment (jmpesp) [reply schema/crdb/crucible-ref-count-records/up08.sql:9]: See above comment.

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 13]
```diff
@@ -91,19 +95,187 @@ enum VolumeGetError {
...
+        volume: Volume,
+        crucible_targets: CrucibleTargets,
+    ) -> Result<Volume, diesel::result::Error> {
         use db::schema::volume::dsl;
 
-        #[derive(Debug, thiserror::Error)]
-        enum VolumeCreationError {
-            #[error("Error from Volume creation: {0}")]
-            Public(Error),
+        let maybe_volume: Option<Volume> = dsl::volume
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs:122]: Not here, no - if the caller is trying to call `volume_create` with a volume that has the same ID as one that already exists or is soft-deleted, then we should disallow that.

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 14]
```diff
@@ -824,9 +1078,29 @@ impl DataStore {
...
+                .filter(
+                    ru_dsl::usage_type
+                        .eq(VolumeResourceUsageType::ReadOnlyRegion),
+                )
+                .filter(ru_dsl::region_id.eq(region.id()))
+                .count()
+                .get_result_async::<i64>(conn)
+                .await?;
+
+            if region_usage_left == 0 {
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs]: Great catch! d4de663b7 adds the `deleting` flag to `region`, which is set in the same way as it is for region_snapshot, preventing this problem. There are also two unit tests checking that this won't happen, along with validation of some volume related invariants that only runs during tests.

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 15]
```diff
@@ -934,52 +1215,237 @@ impl DataStore {
...
-                .left_join(
-                    region_snapshot_dsl::region_snapshot
-                        .on(region_snapshot_dsl::region_id.eq(region_dsl::id)),
-                )
-                .filter(
-                    region_snapshot_dsl::volume_references
-                        .eq(0)
-                        .or(region_snapshot_dsl::volume_references.is_null()),
+        let mut regions: Vec<Uuid> =
+            Vec::with_capacity(3 * num_read_write_subvolumes);
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs]: yes - 1b57804b9

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 16]
```diff
@@ -934,52 +1215,237 @@ impl DataStore {
...
+
+        let mut region_snapshots: Vec<RegionSnapshotV3> =
+            Vec::with_capacity(crucible_targets.read_only_targets.len());
+
+        // First, grab read-write regions - they're not shared, but they are
+        // not candidates for deletion if there are region snapshots
+        {
+            let mut read_write_targets =
+                Vec::with_capacity(3 * num_read_write_subvolumes);
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs]: same - 1b57804b9

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 17]
```diff
@@ -91,19 +95,187 @@ enum VolumeGetError {
...
+            }
+        }
+
+        Ok(volume)
+    }
+
+    async fn target_to_region(
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        err: &OptionalError<AddrParseError>,
+        target: &str,
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs]: I agree, and changed this in 0e0a63951

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 18]
```diff
@@ -1895,6 +2426,8 @@ pub struct VolumeReplacementParams {
 #[derive(Debug, Clone, Copy)]
 pub struct VolumeWithTarget(pub Uuid);
 
+// Note: it would be easier to pass around strings, but comparison could fail
+// due to formatting issues, so pass around SocketAddrV6 here
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs]: see 0e0a63951

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 19]
```diff
@@ -2409,24 +2956,188 @@ impl DataStore {
...
+                                )
+                            ))
+                        );
+                    };
+
+                    // This function may be called with a replacement volume
+                    // that is completely blank, to be filled in later by this
+                    // function. `volume_create` will have been called but will
+                    // not have added any volume resource usage records, because
+                    // it was blank!
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs]: No, this was bad wording! Changed in 58169feb6, let me know if that reads better

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 20]
```diff
@@ -2409,24 +2956,188 @@ impl DataStore {
...
+                    //
+                    // The indention leaving this transaction is that the
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs]: also 58169feb6

File: schema/crdb/crucible-ref-count-records/up08.sql (hunk 1 above)
Comment (jmpesp) [reply schema/crdb/crucible-ref-count-records/up08.sql:9]: done in 58169feb6 (plus a comment)

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 21]
```diff
@@ -2508,6 +3213,65 @@ pub fn read_only_resources_associated_with_volume(
...
+            sub_volumes.len()
+        }
+
+        VolumeConstructionRequest::Url { .. } => 0,
+
+        VolumeConstructionRequest::Region { .. } => {
+            // We don't support a pure Region VCR at the volume
+            // level in the database, so this choice should
+            // never be encountered.
+            panic!("Region not supported as a top level volume");
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs]: yeah, removed in 3849fe856

File: nexus/db-queries/src/db/datastore/volume.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs]: This change found incorrect ipv6 addresses in the tests! b8cccb2c5

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 22]
```diff
@@ -91,19 +95,187 @@ enum VolumeGetError {
...
+    #[error("Address parsing error during Volume creation: {0}")]
+    AddressParseError(#[from] AddrParseError),
+
+    #[error("Could not match read-only resource to {0}")]
+    CouldNotFindResource(String),
+}
+
 impl DataStore {
-    pub async fn volume_create(&self, volume: Volume) -> CreateResult<Volume> {
+    async fn volume_create_txn(
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs]: yes, changed in ce56ca3e7

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 23]
```diff
@@ -1197,20 +1675,73 @@ impl DataStore {
...
+                                        &rop_vcr,
+                                        &mut crucible_targets,
+                                    );
+                                    crucible_targets
+                                };
+
+                                for read_only_target in crucible_targets.read_only_targets {
+                                    let sub_err = OptionalError::new();
+
+                                    let maybe_usage = Self::read_only_target_to_volume_resource_usage(
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs]: ce56ca3e7 reflows the file so nothing is longer than 80

File: schema/crdb/crucible-ref-count-records/up08.sql (hunk 1 above)
Comment (jmpesp) [reply schema/crdb/crucible-ref-count-records/up08.sql:9]: I am leaving a comment so that Github stops bringing me to this text box. Please ignore!

File: nexus/db-model/src/volume_resource_usage.rs [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+#[derive(Debug, Clone)]
+pub enum VolumeResourceUsage {
+    ReadOnlyRegion { region_id: Uuid },
+
+    RegionSnapshot { dataset_id: Uuid, region_id: Uuid, snapshot_id: Uuid },
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-model/src/volume_resource_usage.rs:66]: I'm working on this but as a separate PR that I'll stage on top of this one

**Comment by jmpesp** [comment]
@leftwo @smklein I'm all done responding to comments, please have another look!

File: nexus/db-queries/src/db/datastore/region.rs [hunk 25]
```diff
@@ -362,6 +362,12 @@ impl DataStore {
                             )
                             .execute_async(&conn).await?;
                     }
+
+                    // Whenever a region is hard-deleted, validate invariants
+                    // for all volumes
+                    #[cfg(any(test, feature = "testing"))]
+                    Self::validate_volume_invariants(&conn).await?;
```
Comment (leftwo) [inline nexus/db-queries/src/db/datastore/region.rs:369]: The `feature = "testing"` means we only do this check during tests?

**Review by leftwo — APPROVED** [review]
Some questions, but I like it all

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 26]
```diff
@@ -2549,184 +2707,184 @@ impl DataStore {
...
+                    return Err(
+                        err.bail(ReplaceRegionError::RegionReplacementError(e))
+                    );
+                }
+            };
 
-                        return Ok(VolumeReplaceResult::ExistingVolumeDeleted);
-                    };
+        if !old_region_in_vcr && new_region_in_vcr {
+            // It does seem like the replacement happened
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs:2774]: You mean

```rust
        if !old_region_in_vcr && new_region_in_vcr {
            // It does seem like the replacement happened
            return Ok(VolumeReplaceResult::AlreadyHappened);
        } else if old_region_in_vcr && !new_region_in_vcr {
            // Replacement hasn't happened yet
        } else if old_region_in_vcr && new_region_in_vcr {
            // XXX wat
        } else if !old_region_in_vcr && !new_region_in_vcr {
            // XXX wat
        }
```
?

File: nexus/db-queries/src/db/datastore/volume.rs [hunk 27]
```diff
@@ -3265,11 +3363,11 @@ pub fn count_read_write_sub_volumes(vcr: &VolumeConstructionRequest) -> usize {
             // We don't support a pure Region VCR at the volume
             // level in the database, so this choice should
             // never be encountered.
-            panic!("Region not supported as a top level volume");
+            bail!("Region not supported as a top level volume");
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs:3366]: These functions had an undocumented panic and instead of documenting it, it should return a 500. Nexus shouldn't panic here, as a badly constructed volume could be accidentally created and using it would always result in a panic.

File: nexus/db-queries/src/db/datastore/volume.rs (hunk 26 above)
[the remark this answers is not in the document]
Comment (leftwo) [reply nexus/db-queries/src/db/datastore/volume.rs:2774]: Yes.

It seems like `old_region_in_vcr && !new_region_in_vcr ` is what we expect, and the other two are things that should not happen?

[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs:2774]: Right, yeah. `old_region_in_vcr && new_region_in_vcr` means we performed a replacement (because new is there) but old is still there, meaning that the replacement couldn't have taken place, and `!old_region_in_vcr && !new_region_in_vcr` means that the thing we're being asked to replace isn't there to begin with, and a replacement hasn't taken place (because new isn't there).

Do you think I should have those cases in the code?

[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/db-queries/src/db/datastore/volume.rs]: We chatted and I did put some comments in the code in f52fa636f

File: schema/crdb/crucible-ref-count-records/up08.sql [hunk 28]
```diff
@@ -0,0 +1,31 @@
...
+   snapshot records to populate the volume resource usage records */
+INSERT INTO volume_resource_usage (
+  SELECT
+    (
+      OVERLAY(
+       OVERLAY(
+        MD5(volume.id::TEXT || dataset_id::TEXT || region_id::TEXT || snapshot_id::TEXT || snapshot_addr || volume_references::TEXT)
+        PLACING '4' from 13
+       )
+       PLACING '8' from 17
```
Comment (smklein) [inline schema/crdb/crucible-ref-count-records/up08.sql:11]: This still feels somewhat dangerous to me, and I'm not sure we have tests for the validity nor uniqueness of the UUIDs being produced by this transition.

The goal here is to produce a `volume_resource_usage` record for each `volume INNER JOIN region_snapshot` that isn't being deleted, right? Couldn't you do this by doing:

```sql
-- End goal: Find all volume/region_snapshot combos that **don't** have a volume_resource_usage record
(volume INNER JOIN region_snapshot) LEFT JOIN (volume_resource_usage)
  ON
(
  volume_resource_usage.volume = volume.id AND
  volume_resource_usage.region_snapshot_region_id = region_snapshot.region_id
)
```

And then write a `WHERE` clause to ensure that  `volume_resource_usage.id` is NULL for these cases?

Basically:

```sql
INSERT INTO  volume_resource_usage
-- <New records, using normal UUID generation>
FROM
-- <The clause above - `volume JOIN region_snapshot`, but also `JOIN`-ing with `volume_resource_usage` to check a record doesn't already exist>
WHERE
volume.time_deleted IS NULL AND
region_snapshot.deleting = false AND
volume_resource_usage.id IS NULL
```

This would still be idempotent! Re-running it should be a no-op on the second iteration

Comment (jmpesp) [reply schema/crdb/crucible-ref-count-records/up08.sql]: This does work, see 2b511b17d, thanks :) The `test_migrate_.*` tests confirm that it works.

**Review by smklein — APPROVED** [review]
